File: scripts/mppi_node.py
#!/usr/bin/python3

import copy
import numpy as np
from scipy.ndimage import distance_transform_edt, binary_dilation

import rclpy
import tf2_ros
from rclpy.node import Node
from message_filters import Subscriber, ApproximateTimeSynchronizer
from rclpy.duration import Duration

from ackermann_msgs.msg import AckermannDriveStamped
from nav_msgs.msg import OccupancyGrid
from visualization_msgs.msg import Marker, MarkerArray
from nav_msgs.msg import Odometry
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Point
from transforms3d.euler import quat2euler

from f1tenth_mppi.utils import *
from f1tenth_mppi.dynamics_models import KBM

class MPPI(Node):
    def __init__(self):
        super().__init__('MPPI')

        # Setup loggers
        self.info_log = rclpy.logging.get_logger('info')
        self.warn_log = rclpy.logging.get_logger('warn')
        self.error_log = rclpy.logging.get_logger('error')

        # Initialize Parameters
        self.initialize_parameters()
        self.u_mean = np.array([self.get_parameter("min_throttle").value, 0.0]) # Mean for sampling trajectories
        
        if self.get_parameter("drive_topic").value == None:
            self.error_log.error("No parameters set, use --ros-args --params-file <FILE>")
            exit()

        # Initialize Dynamics Model
        self.model = KBM(self.get_parameter("wheelbase").value,
                         self.get_parameter("min_throttle").value,
                         self.get_parameter("max_throttle").value,
                         self.get_parameter("max_steer").value,
                         self.get_parameter("dt").value)

        # Initialize Occupancy Grid
        self.og = OccupancyGrid() # Create your occupancy grid here
        self.og.header.frame_id = self.get_parameter("vehicle_frame").value
        self.og.info.resolution = self.get_parameter("cost_map_res").value
        self.og.info.width = self.get_parameter("cost_map_width").value
        self.og.info.height = self.get_parameter("cost_map_width").value
        self.og.info.origin.position.x = 0.0
        self.og.info.origin.position.y = -(self.og.info.height * self.og.info.resolution) / 2

        # Initialize Cost Map
        self.cost_map = copy.deepcopy(self.og)

        # Load Waypoints
        try:
            self.waypoints = load_waypoints(self.get_parameter("waypoint_path").value)
        except Exception as e:
            self.error_log.error("Issue loading waypoints")
            self.error_log.error(e)
            exit()

        # Setup the TF2 buffer and listener to capture transforms
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)

        # Create Publishers
        self.drive_pub_ = self.create_publisher(AckermannDriveStamped,
                                                self.get_parameter("drive_topic").value,
                                                10)
        self.occupancy_pub_ = self.create_publisher(OccupancyGrid,
                                                self.get_parameter("occupancy_topic").value,
                                                10)
        self.cost_map_pub_ = self.create_publisher(OccupancyGrid,
                                                self.get_parameter("cost_map_topic").value,
                                                10)
        self.marker_pub_ = self.create_publisher(MarkerArray,
                                                self.get_parameter("marker_topic").value,
                                                10)
        
        # Visualize waypoints
        if self.get_parameter("visualize").value:
            self.publish_markers(self.waypoints, np.array([1.0, 0.0, 0.0]), "waypoints")

        # Create message_filters Subscribers
        self.scan_sub_ = Subscriber(self, LaserScan, self.get_parameter("scan_topic").value)
        self.pose_sub_ = Subscriber(self, Odometry, self.get_parameter("pose_topic").value)

        # Create time synchronized callback
        self.time_sync = ApproximateTimeSynchronizer([self.scan_sub_, self.pose_sub_], 10, 0.1)
        self.time_sync.registerCallback(self.callback)

        self.info_log.info("MPPI node initialized")

    def initialize_parameters(self):
        '''
        Initialize all parameters as None so parameters can be loaded from yaml
        '''
        self.declare_parameters(
            namespace='',
            parameters=[
                ('visualize', None),
                ('waypoint_path', None),
                ('vehicle_frame', None),
                ('drive_topic', None),
                ('occupancy_topic', None),
                ('cost_map_topic', None),
                ('marker_topic', None),
                ('pose_topic', None),
                ('scan_topic', None),
                ('wheelbase', None),
                ('min_throttle', None),
                ('max_throttle', None),
                ('max_steer', None),
                ('dt', None),
                ('num_trajectories', None),
                ('steps_trajectories', None),
                ('v_sigma', None),
                ('omega_sigma', None),
                ('cost_map_width', None),
                ('cost_map_res', None),
                ('obstacle_weight', None),
                ('raceline_weight', None),
                ('obstacle_dilation', None),
                ('raceline_dilation', None),
                ('heading_weight', None),
            ])
        
    def callback(self, scan_msg: LaserScan, pose_msg: Odometry):
        '''
        Time synchronized callback to handle LaserScan and Odometry messages

        Args:
            scan_msg (LaserScan): LaserScan data from the LiDAR
            pose_msg (Odometry): Odometry message from the particle filter
        '''

        self.info_log.info("Recieved scan_msg and pose_msg")

        # Create Occupancy Grid
        self.info_log.info("Creating occupancy grid")
        occupancy_grid = self.create_occupancy_grid(scan_msg)
        
        # Create Cost Map
        self.info_log.info("Creating cost map")
        try:
            cost_map = self.update_cost_map(occupancy_grid)
        except Exception as e:
            self.warn_log.warn("Error updating cost map, skipping iteration")
            print(e)
            return

        # Create Trajectories
        self.info_log.info("Sampling trajectories")
        trajectories, actions = self.sample_trajectories(self.get_parameter("num_trajectories").value,
                                                         self.get_parameter("steps_trajectories").value)

        # Evaluate Trajectories
        self.info_log.info("Evaluating trajectories")
        min_cost_idx = self.evaluate_trajectories(cost_map, trajectories, occupancy_grid)

        if min_cost_idx == -1:
            drive_msg = AckermannDriveStamped()
            drive_msg.drive.steering_angle = actions[min_cost_idx, 0, 1]
            drive_msg.drive.speed = 0.0
            self.drive_pub_.publish(drive_msg)
            return

        # # Update u_mean
        # self.u_mean = actions[min_cost_idx, 0]
        self.u_mean[0] = actions[min_cost_idx, 0, 0]

        # Publish AckermannDriveStamped Message
        self.info_log.info("Publishing drive command")
        drive_msg = AckermannDriveStamped()
        drive_msg.drive.steering_angle = actions[min_cost_idx, 0, 1]
        drive_msg.drive.speed = actions[min_cost_idx, 0, 0]
        self.drive_pub_.publish(drive_msg)
    
    def create_occupancy_grid(self, scan_msg: LaserScan) -> np.ndarray:
        '''
        Process the LaserScan data into an occupancy grid

        Args:
            scan_msg (LaserScan): LaserScan data from the LiDAR
        Returns:
            occupancy_grid (ndarray): The processed occupancy grid
        '''

        # Initialize an empty occupancy grid (2D array)
        occupancy_grid = np.zeros((self.og.info.width, self.og.info.width), dtype=int)

        # Convert laser scan ranges to grid coordinates
        angles = scan_msg.angle_min + scan_msg.angle_increment * np.arange(len(scan_msg.ranges))
        ranges = np.array(scan_msg.ranges)

        x_coords = np.round((ranges * np.sin(angles)) / self.og.info.resolution + self.og.info.width / 2).astype(int)
        y_coords = np.round((ranges * np.cos(angles)) / self.og.info.resolution).astype(int)

        # Filter out any points that fall outside the grid boundaries
        valid_mask = ((x_coords > 0) & (x_coords < self.og.info.width) &
                      (y_coords > 0) & (y_coords < self.og.info.height))

        x_coords = x_coords[valid_mask]
        y_coords = y_coords[valid_mask]

        # Mark occupied cells in the grid
        occupancy_grid[x_coords, y_coords] = 100

        # Define a simple structuring element (kernel) for dilation
        # dilation_kernel = np.array([[0, 1, 0],
        #                             [1, 1, 1],
        #                             [0, 1, 0]], dtype=bool)

        occupancy_grid[45:55, 0:10] = 0 # TODO: Put this in real world coordinates

        # Apply binary dilation to expand obstacles
        dilation_kernel = np.ones((self.get_parameter("obstacle_dilation").value, self.get_parameter("obstacle_dilation").value), dtype=bool)
        occupancy_grid = binary_dilation(occupancy_grid, structure=dilation_kernel).astype(int) * 100

        # Prepare and publish the updated occupancy grid
        if self.get_parameter("visualize").value:
            self.og.data = occupancy_grid.flatten().tolist()
            self.og.header.stamp = self.get_clock().now().to_msg()
            self.occupancy_pub_.publish(self.og)

        return occupancy_grid
    
    def update_cost_map(self, occupancy_grid: np.ndarray) -> np.ndarray:
        '''
        Update cost map based on current environment.
        We want areas that deviate from the raceline to have higher cost.
        Areas that are closer to obstacles in the occupancy_grid should have higher cost.

        Args:
            occupancy_grid (ndarray): The processed occupancy grid
        Returns:
            cost_map (ndarray): The cost map
        '''

        # Get transform from map to ego frame
        transform = self.tf_buffer.lookup_transform(self.get_parameter("vehicle_frame").value,
                                                    "map", rclpy.time.Time(), timeout=Duration(seconds=0.1))

        # Get vehicle yaw
        qx = transform.transform.rotation.x
        qy = transform.transform.rotation.y
        qz = transform.transform.rotation.z
        qw = transform.transform.rotation.w
        yaw = np.arctan2(2 * (qw * qz + qx * qy), 1 - 2 * (qy**2 + qz**2))

        # Determine translation and rotation
        R = np.array([[np.cos(yaw), -np.sin(yaw)],
                    [np.sin(yaw),  np.cos(yaw)]])
        T = np.array([transform.transform.translation.x,
                      transform.transform.translation.y])

        # Transform all waypoints from map to ego frame
        points_transformed = np.dot(self.waypoints[:, :2], R.T) + T

        # Convert the transformed points into grid indices.
        x_idx = np.round((points_transformed[:, 0] / self.cost_map.info.resolution)).astype(int)
        y_idx = np.round((points_transformed[:, 1] / self.cost_map.info.resolution) + self.cost_map.info.height / 2).astype(int)

        raceline_mask = np.zeros_like(occupancy_grid, dtype=int)
        mask_shape = raceline_mask.shape
        valid = (x_idx >= 0) & (x_idx < mask_shape[0]) & (y_idx >= 0) & (y_idx < mask_shape[1])
        raceline_mask[y_idx[valid], x_idx[valid]] = 100

        # Apply binary dilation to expand raceline
        # dilation_kernel = np.array([[0, 1, 0],
        #                             [1, 1, 1],
        #                             [0, 1, 0]], dtype=bool)
        dilation_kernel = np.ones((self.get_parameter("raceline_dilation").value, self.get_parameter("raceline_dilation").value), dtype=bool)
        raceline_mask = binary_dilation(raceline_mask, structure=dilation_kernel).astype(int) * 100

        # Compute the distance from raceline, for each grid cell
        raceline_cost = distance_transform_edt(raceline_mask == 0)

        # Final cost map is a weighted sum of the obstacle cost and raceline cost
        cost_map = self.get_parameter("raceline_weight").value * raceline_cost
        # cost_map = 100 * (cost_map / cost_map.max()) # normalize so that it's within 0-100
        # cost_map = np.clip(cost_map, 0, 100).astype(int) # NOTE: Should we be clipping or normalizing? Do we need to?

        if self.get_parameter("visualize").value:
            # normalize so that it's within 0-100
            cost_map_vis = 100 * (cost_map / cost_map.max())
            self.cost_map.data = cost_map_vis.astype(int).flatten().tolist()
            self.cost_map.header.stamp = self.get_clock().now().to_msg()
            self.cost_map_pub_.publish(self.cost_map)

        return cost_map / cost_map.max()

    def sample_trajectories(self, num_trajectories: int, steps_trajectories: int):
        '''
        Sample random actions from distribution and generate trajectories using model

        Args:
            num_trajectories (int): The number of trajectories to sample
            steps_trajectories (int): The number of steps for each trajectory
        Returns:
            trajectories (ndarray): (num_trajectories x steps_trajectories x 3) array of trajectories
            actions (ndarray): (num_trajectories x steps_trajectories - 1) array of actions
        '''

        # Sample control values
        # v = self.u_mean[0] + np.random.randn(num_trajectories, steps_trajectories - 1, 1) * self.get_parameter("v_sigma").value
        # omega = self.u_mean[1] + np.random.randn(num_trajectories, steps_trajectories - 1, 1) * self.get_parameter("omega_sigma").value

        v = (self.get_parameter("max_throttle").value + np.random.randn(num_trajectories, 1, 1) * self.get_parameter("v_sigma").value)
        omega = (self.u_mean[1] + np.random.randn(num_trajectories, 1, 1) * self.get_parameter("omega_sigma").value)

        v = np.repeat(v, steps_trajectories - 1, axis=1)
        omega = np.repeat(omega, steps_trajectories - 1, axis=1)

        # Limit control values
        v = np.clip(v, self.get_parameter("min_throttle").value, self.get_parameter("max_throttle").value)
        omega = np.clip(omega, -self.get_parameter("max_steer").value, self.get_parameter("max_steer").value)

        actions = np.concatenate((v, omega), axis=2)

        # Sample trajectories
        trajectories = np.zeros((num_trajectories, steps_trajectories, 3))
        for i in range(steps_trajectories - 1):
            trajectories[:, i + 1] = self.model.predict_euler(trajectories[:, i], actions[:, i])

        # Publish a subset of trajectories
        self.publish_trajectories(trajectories[:5])

        return trajectories, actions
    
    def evaluate_trajectories(self, cost_map: np.ndarray, trajectories: np.ndarray, occupancy_grid: np.ndarray) -> int:
        '''
        Evaluate trajectories using the cost map

        Args:
            cost_map (ndarray): The cost map
            trajectories (np.ndarray): (num_trajectories x steps_trajectories x 3) Sampled trajectories
            occupancy_grid (ndarray): The processed occupancy grid
        Returns:
            min_cost_idx (int): The index of the trajectory with the lowest cost
        '''

        # Convert each trajectory to the cost map frame
        trajectories_pixels = trajectories / self.cost_map.info.resolution
        trajectories_pixels[:, :, 0] = trajectories_pixels[:, :, 0]
        trajectories_pixels[:, :, 1] = trajectories_pixels[:, :, 1] + self.cost_map.info.height/2

        # Handle trajectories that fall outside of the cost map
        trajectories_pixels = np.clip(trajectories_pixels[:, :, :2], 0, self.cost_map.info.width - 1)

        # if a trajectory touches an obstacle, then just completely ignore it 
        # print(f"shape of traj pixels is {trajectories_pixels.shape}") # NxTx3
        # print(f"Shape of occupancy grid is {occupancy_grid.shape}")

        # bad_traj = [1, 0, 1, 0, 0,.. N]
        check_obs = occupancy_grid.astype(bool)[trajectories_pixels[:, :, 1].astype(int), trajectories_pixels[:, :, 0].astype(int)] # NxT
        bad_trajs = np.any(check_obs==True, axis=1) # (N,)

        for i in range(check_obs.shape[0]):
            try:
                idx = np.where(check_obs[i] == True)[0][0]
                check_obs[i, idx:] = True
            except:
                pass

        # Compute each trajectory's position score and normalize
        traj_scores = np.sum(cost_map[trajectories_pixels[:, :, 1].astype(int), trajectories_pixels[:, :, 0].astype(int)], axis=1)
        traj_scores /= traj_scores.max()

        traj_scores[bad_trajs] = 50
        # if np.all(traj_scores == np.inf):
        #     return -1
        min_cost_idx = np.argmin(traj_scores)

        # Publish a lowest cost trajectory
        self.publish_trajectories(np.expand_dims(trajectories[min_cost_idx], 0), np.array([1.0, 0.0, 0.0]))

        return min_cost_idx

    def publish_trajectories(self, points: np.ndarray, color: np.ndarray = np.array([0.0, 0.0, 1.0])):
        '''
        Publish MarkerArray message of lines

        Args:
            points (ndarray): NxMx2 array of points to publish
            color (ndarray): The color of the points
        '''

        if not self.get_parameter("visualize").value:
            return

        # Generate MarkerArray message
        marker_array = MarkerArray()
        for j in range(points.shape[0]):
            for i in range(points.shape[1] - 1):
                marker = Marker()
                marker.header.frame_id = "ego_racecar/laser"
                marker.id = j * points.shape[1] + i
                marker.header.stamp = self.get_clock().now().to_msg()
                marker.type = Marker.LINE_LIST
                marker.action = Marker.ADD
                marker.color.r = color[0]
                marker.color.g = color[1]
                marker.color.b = color[2]
                marker.color.a = 1.0
                marker.scale.x = 0.01

                p1 = Point()
                p1.x = points[j, i, 0]
                p1.y = points[j, i, 1]
                p2 = Point()
                p2.x = points[j, i+1, 0]
                p2.y = points[j, i+1, 1]

                marker.points = [p1, p2]
                marker_array.markers.append(marker)

        # Publish MarkerArray Message
        self.marker_pub_.publish(marker_array)

    def publish_markers(self, points: np.ndarray, color: np.ndarray = np.array([1.0, 0.0, 0.0]), ns: str = ""):
        '''
        Publish MarkerArray message of points

        Args:
            points (ndarray): Nx2 array of points to publish
            color (ndarray): The color of the points
        '''

        if not self.get_parameter("visualize").value:
            return
        
        # Generate MarkerArray message
        marker_array = MarkerArray()
        for i in range(len(points)):
            marker = Marker()
            marker.ns = ns
            marker.header.frame_id = "map"
            marker.id = i + 1
            marker.header.stamp = self.get_clock().now().to_msg()
            marker.type = Marker.SPHERE
            marker.action = Marker.ADD
            marker.pose.position.x = points[i, 0]
            marker.pose.position.y = points[i, 1]
            marker.color.r = color[0]
            marker.color.g = color[1]
            marker.color.b = color[2]
            marker.color.a = 1.0
            marker.scale.x = 0.1
            marker.scale.y = 0.1
            marker.scale.z = 0.1

            marker_array.markers.append(marker)

        # Publish MarkerArray Message
        self.marker_pub_.publish(marker_array)

def main(args=None):
    rclpy.init(args=args)
    mppi_node = MPPI()
    rclpy.spin(mppi_node)

    mppi_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/mppi_node.cpp
#include <sstream>
#include <string>
#include <cmath>
#include <vector>
#include <iostream>
#include <fstream> 
#include <stdexcept>
#include <random>

#include "rclcpp/rclcpp.hpp"
#include "utils.hpp"
#include "dynamics_models.hpp"

#include <message_filters/subscriber.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <message_filters/synchronizer.h>

#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include <visualization_msgs/msg/marker_array.hpp>
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"

#include "geometry_msgs/msg/point_stamped.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/transform_broadcaster.h"
#include "tf2_ros/buffer.h"
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include "ament_index_cpp/get_package_prefix.hpp"
#include <visualization_msgs/msg/marker.hpp>

/// CHECK: include needed ROS msg type headers and libraries

using std::placeholders::_1;
using std::placeholders::_2;
#define rad2deg 180.0 / M_PI
#define deg2rad M_PI / 180

class MPPI : public rclcpp::Node
{

private:
    // Publishers
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_pub_;
    rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr occupancy_pub;
    rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_pub_;

    // Subscribers
    message_filters::Subscriber<sensor_msgs::msg::LaserScan> scan_sub_;
    message_filters::Subscriber<nav_msgs::msg::Odometry> pose_sub_;
    typedef message_filters::sync_policies::ApproximateTime<
        sensor_msgs::msg::LaserScan, nav_msgs::msg::Odometry> SyncPolicy;
    typedef message_filters::Synchronizer<SyncPolicy> Sync;
    std::shared_ptr<Sync> sync_;

    // Transform objects
    std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_{nullptr};
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;

    // Other
    std::vector<std::vector<double>> waypoints;
    KBM model;
    nav_msgs::msg::OccupancyGrid og = nav_msgs::msg::OccupancyGrid();
    std::vector<std::vector<double>> u_prev;

public:
    MPPI() : Node("mppi_node") {
        /// TODO: Setup Loggers

        // Initialize Parameters
        initialize_parameters();

        // Initialize Model
        model.set_params(this->get_parameter("wheelbase").as_double(),
                         this->get_parameter("dt").as_double());

        // Initialize Occupancy Grid
        og.header.frame_id = this->get_parameter("vehicle_frame").as_string();
        og.info.resolution = this->get_parameter("cost_map_res").as_double();
        og.info.width = this->get_parameter("cost_map_width").as_int();
        og.info.height = this->get_parameter("cost_map_width").as_int();
        og.info.origin.position.x = 0.0;
        og.info.origin.position.y = -og.info.height * og.info.resolution / 2;

        // Load Waypoints
        try {
            waypoints = load_waypoints(this->get_parameter("waypoint_path").as_string());
            /// TODO: Normalize speed to min + max
        } catch (const std::runtime_error& error) {
            std::cout << "Issue loading waypoints\n";
            std::cerr << error.what() << std::endl;
        }
        
        // Initialize Previous Control Input Sequence
        for (int i = 0; i < this->get_parameter("steps_trajectories").as_int(); i++) {
            std::vector<double> zeros = {0.0, 0.0};
            u_prev.push_back(zeros);
        }

        // Setup TF2 Buffer + Listener
        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
        tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

        // Create Publishers
        drive_pub_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(this->get_parameter("drive_topic").as_string(), 10);
        occupancy_pub = this->create_publisher<nav_msgs::msg::OccupancyGrid>(this->get_parameter("occupancy_topic").as_string(), 10);
        marker_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>(this->get_parameter("marker_topic").as_string(), 10);

        // Visualize Waypoints
        publish_markers(marker_pub_, waypoints);

        // Create message_filter Subscribers
        scan_sub_.subscribe(this, this->get_parameter("scan_topic").as_string());
        pose_sub_.subscribe(this, this->get_parameter("pose_topic").as_string());

        // Create time synchronized callback
        sync_ = std::make_shared<Sync>(SyncPolicy(10), scan_sub_, pose_sub_);
        sync_->registerCallback(std::bind(&MPPI::callback, this, _1, _2));
    }

    void initialize_parameters() {
        this->declare_parameter("visualize", false);
        this->declare_parameter("waypoint_path", "aims_lockers.csv");
        this->declare_parameter("vehicle_frame", "ego_racecar/laser");
        this->declare_parameter("drive_topic", "/drive");
        this->declare_parameter("occupancy_topic", "/occupancy");
        this->declare_parameter("marker_topic", "/trajectory");
        this->declare_parameter("pose_topic", "/ego_racecar/odom");
        this->declare_parameter("scan_topic", "/scan");
        this->declare_parameter("wheelbase", 0.33);
        this->declare_parameter("min_throttle", 0.5);
        this->declare_parameter("max_throttle", 1.5);
        this->declare_parameter("max_steer", 0.4139);
        this->declare_parameter("dt", 0.1);
        this->declare_parameter("num_trajectories", 50);
        this->declare_parameter("steps_trajectories", 25);
        this->declare_parameter("v_sigma", 0.05);
        this->declare_parameter("omega_sigma", 0.2);
        this->declare_parameter("lambda", 10.0);
        this->declare_parameter("cost_map_width", 100);
        this->declare_parameter("cost_map_res", 0.05);
        this->declare_parameter("occupancy_dilation", 10);
    }

    void callback(const std::shared_ptr<const sensor_msgs::msg::LaserScan> scan_msg,
                  const std::shared_ptr<const nav_msgs::msg::Odometry> pose_msg)
    {
        /// TODO: Create Occupancy Grid
        std::cout << "HERE" << std::endl;

        // Get random seed
        unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
        std::default_random_engine generator(seed);

        // ---------- MPPI ----------
        // Set Initial State
        tf2::Quaternion q(
            pose_msg->pose.pose.orientation.x,
            pose_msg->pose.pose.orientation.y,
            pose_msg->pose.pose.orientation.z,
            pose_msg->pose.pose.orientation.w
        );
        
        // Convert quaternion to roll, pitch, yaw
        double roll, pitch, yaw;
        tf2::Matrix3x3(q).getRPY(roll, pitch, yaw);

        if (yaw < 0) {yaw += 2 * M_PI;}

        std::vector<double> x0 = {pose_msg->pose.pose.position.x,
                                  pose_msg->pose.pose.position.y,
                                  yaw};

        // Load Previous Control Input Sequence
        std::vector<std::vector<double>> u = u_prev;
        
        // Initialize Noise
        std::normal_distribution<double> v_noise(0.0, this->get_parameter("v_sigma").as_double());
        std::normal_distribution<double> omega_noise(0.0, this->get_parameter("omega_sigma").as_double());
        std::vector<std::vector<std::vector<double>>> epsilon; // num_trajectories x steps_trajectories x 2

        // Initialize State Cost
        std::vector<double> S;

        // Loop through timesteps
        for (int i = 0; i < this->get_parameter("num_trajectories").as_int(); i++) {
            // Reset
            std::vector<double> x = x0;
            std::vector<std::vector<double>> epsilon_i;
            double v;
            S.push_back(0.0);
            for (int j = 1; j <= this->get_parameter("steps_trajectories").as_int(); j++) {
                // Add noise
                v = clamp(u[j - 1][0] + v_noise(generator), this->get_parameter("min_throttle").as_double(), this->get_parameter("max_throttle").as_double());
                double omega = clamp(u[j - 1][1] + omega_noise(generator), -this->get_parameter("max_steer").as_double(), this->get_parameter("max_steer").as_double());
                epsilon_i.push_back({v - u[j - 1][0], omega - u[j - 1][1]});

                // Update State
                x = model.predict_euler(x, {v, omega});

                // Add Stage Cost
                S[i] += compute_cost(x, v);
            }

            // Add terminal Cost
            /// TODO: Add magic number to parameters or modify
            S[i] += compute_cost(x, v) * 10.0;
            
            epsilon.push_back(epsilon_i);
        }

        // Compute information theoretic weights
        std::vector<double> w = compute_weights(S);

        // Calculate Added Noise
        std::vector<std::vector<double>> w_epsilon;
        for (int i = 0; i < this->get_parameter("steps_trajectories").as_int(); i++) {
            w_epsilon.push_back({0.0, 0.0});
            for (int j = 0; j < this->get_parameter("num_trajectories").as_int(); j++) {
                w_epsilon[i][0] += w[j] * epsilon[j][i][0];
                w_epsilon[i][1] += w[j] * epsilon[j][i][1];
            }
        }

        /// TODO: Smooth added noise

        // Update + Clip Control Input Sequence
        for (int i = 0; i < this->get_parameter("steps_trajectories").as_int(); i++) {
            u[i][0] += w_epsilon[i][0];
            u[i][1] += w_epsilon[i][1];

            u[i][0] = clamp(u[i][0], this->get_parameter("min_throttle").as_double(), this->get_parameter("max_throttle").as_double());
            u[i][1] = clamp(u[i][1], -this->get_parameter("max_steer").as_double(), this->get_parameter("max_steer").as_double());
        }

        u_prev = u;
        // --------------------------

        // Publish Trajectory
        std::vector<std::vector<double>> x_traj;
        x_traj.push_back(x0);
        for (int i = 1; i < this->get_parameter("steps_trajectories").as_int(); i++) {
            x_traj.push_back(model.predict_euler(x_traj[i-1], u[i]));
        }
        publish_trajectory(marker_pub_, x_traj);

        // Publish AckermannDriveStamped Message
        auto drive_msg = ackermann_msgs::msg::AckermannDriveStamped();
        drive_msg.header.stamp = this->get_clock()->now();
        drive_msg.drive.speed = u[0][0];
        drive_msg.drive.steering_angle = u[0][1];
        // drive_pub_->publish(drive_msg);
    }

    double compute_cost(std::vector<double> x_t, double v_t) {
        // Set Stage Cost weights
        /// TODO: Add to parameters
        std::vector<double> stage_cost_weights = {13.5, 13.5, 5.5, 5.5};
        
        // Find nearest waypoint
        std::vector<double> x_ref = get_nearest_waypoint(x_t[0], x_t[1]);

        // Fix yaw
        x_ref[3] += M_PI / 2; // Fix waypoint offset
        if (x_ref[2] < 0) {x_ref[2] += 2 * M_PI;} // -2 * M_PI < x_ref[2] < 2 * M_PI
        if (x_ref[2] - x_t[2] > 4.5) {x_ref[2] -= 2 * M_PI;} // Bring x_t[2] & x_ref[2] together
        else if (x_ref[2] - x_t[2] < -4.5) {x_ref[2] += 2 * M_PI;} // Bring x_t[2] & x_ref[2] together

        // Compute stage cost
        double stage_cost = stage_cost_weights[0] * std::pow(x_t[0] - x_ref[0], 2)
                          + stage_cost_weights[1] * std::pow(x_t[1] - x_ref[1], 2)
                          + stage_cost_weights[2] * std::pow(x_t[2] - x_ref[2], 2)
                          + stage_cost_weights[3] * std::pow(v_t - x_ref[3], 2);

        return stage_cost;
    }

    std::vector<double> get_nearest_waypoint(double x, double y) {
        float closest_dist = INFINITY;
        int closest_idx = 0;

        std::vector<double> x_t = {x, y};
        for (unsigned long i = 0; i < waypoints.size(); i++) {
            std::vector<double> point = {waypoints[i][0], waypoints[i][0]};
            float dist = get_dist(x_t, point);

            if (dist < closest_dist) {
                closest_idx = i;
                closest_dist = dist;
            }
        }

        return waypoints[closest_idx];
    }
    
    std::vector<double> compute_weights(std::vector<double> S) {
        double rho = *std::min_element(S.begin(), S.end());

        double eta = 0;
        for (int i = 0; i < this->get_parameter("num_trajectories").as_int(); i++) {
            eta += exp(-1.0 / this->get_parameter("lambda").as_double()) * (S[i] - rho);
        }

        std::vector<double> w;
        for (int i = 0; i < this->get_parameter("num_trajectories").as_int(); i++) {
            w.push_back((1.0 / eta) * exp(-1.0 / this->get_parameter("lambda").as_double()) * (S[i] - rho));
        }

        return w;
    }

    ~MPPI() {}
};
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MPPI>());
    rclcpp::shutdown();
    return 0;
}